MekHQ/src/mekhq/campaign/mission/CrewSkillUpgrader.java

```diff
@@ -85,34 +88,36 @@ public void upgradeCrew(Entity entity) {
         // this sets the weight limit and how many SPAs we can assign
         // this is described in some detail in CamOps page 70 (Special Pilot Abilities)
         int upgradeRoll = Compute.randomInt(4);
-        if(upgradeRoll != 3) {
+        if (upgradeRoll != 3) {
             return;
         }
-        
+
         double skillAvg = (entity.getCrew().getGunnery() + entity.getCrew().getPiloting()) / 2.0;
         double xpCap = 0;
         int spaCap = 0;
-        
+
         // elite
-        if(skillAvg < 3) {
+        if (skillAvg < 3) {
             xpCap = maxAbilityXPCost;
             spaCap = 3;
         // veteran
-        } else if(skillAvg < 4) {
+        } else if (skillAvg < 4) {
             xpCap = twoThirdsXPCost;
             spaCap = 2;
         // regular
-        } else if(skillAvg < 5) {
+        } else if (skillAvg < 5) {
             xpCap = oneThirdXPCost;
             spaCap = 1;
         }
-        
-        // algorithm: 
+
+        // algorithm:
         // we want a maximum # of SPAs, capped, in total at a max XP cost
         // every time we generate an SPA, we reduce the max available XP
         // this logic also prevents attempting to assign an SPA when there are no SPAs
         // that cost less XP than the remaining cap
-        for(int x = 0; (x < spaCap) && (xpCap > minAbilityCost); x++) {
+        // We also want to ensure this only happens a maximum of a thousand times, to prevent an
+        // infinite loop
+        for (int x = 0, y = 0; (x < spaCap) && (xpCap > minAbilityCost) && (y < 1000); x++, y++) {
             int spaCost = addSingleSPA(entity, xpCap);
             // if we didn't assign an SPA, let's reroll it.
             if (spaCost == 0) {
```

Comment:
I think this logic is the main problem, it drops the x counter back one if we fail to generate an SPA. I think it needs to be taken out and the addSingleSPA modified, as described in the comment there.